guacamole-ext/src/main/java/org/apache/guacamole/token/TokenFilter.java

```diff
@@ -216,19 +238,71 @@ public String filter(String input) {
 
             // Update last regex match
             endOfLastMatch = tokenMatcher.end();
-            
+
         }
 
         // Append any remaining non-token text
         output.append(input.substring(endOfLastMatch));
-        
+
         return output.toString();
-       
+
+    }
+
+    /**
+     * Filters the given string, replacing any tokens with their corresponding
+     * values. Any tokens present in the given string which lack values will
+     * be interpreted as literals.
+     *
+     * @param input
+     *     The string to filter.
+     *
+     * @return
+     *     A copy of the input string, with any tokens replaced with their
+     *     corresponding values.
+     */
+    public String filter(String input) {
+
+        // Filter with strict mode disabled (should always succeed)
+        try {
+            return filter(input, false);
+        }
+
+        // GuacamoleTokenUndefinedException cannot be thrown when strict mode
+        // is disabled
+        catch (GuacamoleTokenUndefinedException e) {
+            throw new IllegalStateException("filter() threw "
+                    + "GuacamoleTokenUndefinedException despite strict mode "
+                    + "being disabled.");
+        }
+
+    }
+
+    /**
+     * Filters the given string, replacing any tokens with their corresponding
+     * values. If any token in the given string has no defined value within
+     * this TokenFilter, a GuacamoleTokenUndefinedException will be thrown.
+     *
+     * @param input
+     *     The string to filter.
+     *
+     * @return
+     *     A copy of the input string, with any tokens replaced with their
+     *     corresponding values.
+     *
+     * @throws GuacamoleTokenUndefinedException
+     *     If at least one token in the given string has no corresponding
+     *     value.
+     */
+    public String filterStrict(String input)
```

Comment:
Is there rationale for creating this `filterStrict()` method and leaving the actual `filter()` method `private` as noted above?  Seems like it would be okay to just make the `filter()` method public??

Reply:
Yes, except that would then require all callers to `filter()` to catch the possible exception. The behavior of `filter()` historically has been the following:

* If a token has an associated value, substitute that value.
* If a token _does not have_ an associated value, perform no substitution and leave the token as-is.

The new strict implementation is all-or-nothing, substituting all tokens and bailing out if any one token has no corresponding value.

The rationale here is that there are two distinct behaviors and two distinct use-cases for `filter()` vs. `filterStrict()`, and so I think it makes more sense to leave the original `filter()` untouched.